cluster-autoscaler/cloudprovider/huaweicloud/huaweicloud_service_manager.go

```diff
@@ -389,3 +413,129 @@ func (csm *cloudServiceManager) deleteScalingPolicy(opts *huaweicloudsdkasmodel.
 	klog.V(1).Infof("delete scaling policy succeed. policy id: %s", opts.ScalingPolicyId)
 	return nil
 }
+
+func (csm *cloudServiceManager) getScalingGroupByID(groupID string) (*huaweicloudsdkasmodel.ScalingGroups, error) {
+	asClient := csm.getASClientFunc()
+	opts := &huaweicloudsdkasmodel.ShowScalingGroupRequest{
+		ScalingGroupId: groupID,
+	}
+	response, err := asClient.ShowScalingGroup(opts)
+	if err != nil {
+		klog.Errorf("failed to show scaling group info. group: %s, error: %v", groupID, err)
+		return nil, err
+	}
+	if response == nil || response.ScalingGroup == nil {
+		return nil, fmt.Errorf("no scaling group found: %s", groupID)
+	}
+
+	return response.ScalingGroup, nil
+}
+
+func (csm *cloudServiceManager) getScalingGroupConfigByID(groupID, configID string) (*huaweicloudsdkasmodel.ScalingConfiguration, error) {
+	asClient := csm.getASClientFunc()
+	opts := &huaweicloudsdkasmodel.ShowScalingConfigRequest{
+		ScalingConfigurationId: configID,
+	}
+	response, err := asClient.ShowScalingConfig(opts)
+	if err != nil {
+		klog.Errorf("failed to show scaling group config. config id: %s, error: %v", configID, err)
+		return nil, err
+	}
+	if response == nil || response.ScalingConfiguration == nil {
+		return nil, fmt.Errorf("no instance in scaling group: %s", groupID)
+	}
+	return response.ScalingConfiguration, nil
+}
+
+func (csm *cloudServiceManager) listFlavors(az string) (*[]huaweicloudsdkecsmodel.Flavor, error) {
+	ecsClient := csm.getECSClientFunc()
+	opts := &huaweicloudsdkecsmodel.ListFlavorsRequest{
+		AvailabilityZone: &az,
+	}
+	response, err := ecsClient.ListFlavors(opts)
+	if err != nil {
+		klog.Errorf("failed to list flavors. availability zone: %s", az)
+		return nil, err
+	}
+
+	return response.Flavors, nil
+}
+
+func (csm *cloudServiceManager) getAsgTemplate(groupID string) (*asgTemplate, error) {
+	sg, err := csm.getScalingGroupByID(groupID)
+	if err != nil {
+		klog.Errorf("failed to get ASG by id:%s,because of %s", groupID, err.Error())
+		return nil, err
+	}
+
+	configuration, err := csm.getScalingGroupConfigByID(groupID, *sg.ScalingConfigurationId)
+
+	for _, az := range *sg.AvailableZones {
+		flavors, err := csm.listFlavors(az)
+		if err != nil {
+			klog.Errorf("failed to list flavors, available zone is: %s, error: %v", az, err)
+			return nil, err
+		}
+
+		for _, flavor := range *flavors {
+			if !strings.EqualFold(flavor.Name, *configuration.InstanceConfig.FlavorRef) {
+				continue
+			}
+
+			vcpus, _ := strconv.ParseInt(flavor.Vcpus, 10, 64)
+			return &asgTemplate{
+				name: flavor.Name,
+				vcpu: vcpus,
+				ram:  int64(flavor.Ram),
+				zone: az,
+			}, nil
+		}
+	}
+	return nil, nil
+}
+
+func (csm *cloudServiceManager) buildNodeFromTemplate(asgName string, template *asgTemplate) (*apiv1.Node, error) {
+	node := apiv1.Node{}
+	nodeName := fmt.Sprintf("%s-asg-%d", asgName, rand.Int63())
+
+	node.ObjectMeta = metav1.ObjectMeta{
+		Name:     nodeName,
+		SelfLink: fmt.Sprintf("/api/v1/nodes/%s", nodeName),
```

Comment:
`SelfLink` has been deprecated. Not sure if we can ignore this field now.
Fix it or not depend on you.

Reply:
Wow . `Kubernetes will stop propagating this field in 1.20 release`  Since we are still in 1.19+, I assume we can't remove this field now.

Reply:
just leave it for now